TLS doesn’t work for the server

Just found out the following bug while actually testing TLS capabilities of the server beyond using https://testssl.sh: Right after establishing TLS, the connection breaks down with the following error:

>
[PROTOCOL] [35CDAFD3-7750-4477-89D2-6D7883B8B472] [core] S: 250-<myserver> Hello <myserver> [192.168.0.0], Haraka is at your service.
[PROTOCOL] [35CDAFD3-7750-4477-89D2-6D7883B8B472] [core] S: 250-PIPELINING
[PROTOCOL] [35CDAFD3-7750-4477-89D2-6D7883B8B472] [core] S: 250-8BITMIME
[PROTOCOL] [35CDAFD3-7750-4477-89D2-6D7883B8B472] [core] S: 250-SIZE 0
[PROTOCOL] [35CDAFD3-7750-4477-89D2-6D7883B8B472] [core] S: 250-STARTTLS
[PROTOCOL] [35CDAFD3-7750-4477-89D2-6D7883B8B472] [core] S: 250 AUTH CRAM-MD5
[PROTOCOL] [35CDAFD3-7750-4477-89D2-6D7883B8B472] [core] C: STARTTLS state=1
[DEBUG] [35CDAFD3-7750-4477-89D2-6D7883B8B472] [core] running unrecognized_command hooks
[DEBUG] [35CDAFD3-7750-4477-89D2-6D7883B8B472] [core] running unrecognized_command hook in tls plugin
[PROTOCOL] [35CDAFD3-7750-4477-89D2-6D7883B8B472] [core] S: 220 Go ahead.
[DEBUG] [-] [core] Upgrading to TLS
[PROTOCOL] [35CDAFD3-7750-4477-89D2-6D7883B8B472] [core] C: EHLO <myserver> state=1
[DEBUG] [35CDAFD3-7750-4477-89D2-6D7883B8B472] [core] running ehlo hooks
[ERROR] [35CDAFD3-7750-4477-89D2-6D7883B8B472] [core] cmd_ehlo failed: Error: We are already running hooks! Fatal error!
[ERROR] [-] [core] undefinedError: We are already running hooks! Fatal error!0[ 'Error: We are already running hooks! Fatal error!',
  '    at Object.plugins.run_hooks (/usr/lib64/node_modules/Haraka/plugins.js:394:15)',
  '    at /usr/lib64/node_modules/Haraka/connection.js:1267:17',
  '    at Connection.reset_transaction (/usr/lib64/node_modules/Haraka/connection.js:630:17)',
  '    at Connection.cmd_ehlo (/usr/lib64/node_modules/Haraka/connection.js:1264:10)',
  '    at Connection.process_line (/usr/lib64/node_modules/Haraka/connection.js:323:29)',
  '    at Connection._process_data (/usr/lib64/node_modules/Haraka/connection.js:468:18)',
  '    at Connection.process_data (/usr/lib64/node_modules/Haraka/connection.js:378:10)',
  '    at pluggableStream.<anonymous> (/usr/lib64/node_modules/Haraka/connection.js:139:14)',
  '    at emitOne (events.js:96:13)',
  '    at pluggableStream.emit (events.js:188:7)' ]
[…]
[PROTOCOL] [35CDAFD3-7750-4477-89D2-6D7883B8B472] [core] S: 421 Internal Server Error
>

I’m running Node v6.7.0 and programmed the new TLSSocket callback according to the docs by connecting to a ```'secureConnect'``` event. It turns out this event is never emitted, instead the event is still named ```'secure'``` as it was when we still used the ```createSocketPair()``` API.

# Steps to reproduce

Try to actually send a message through STARTTLS. As soon as the client sends its EHLO the server sends the error message ```421 Internal Server Error``` and terminates the connection.

What does work is sending mail without STARTTLS.

--- a/tls_socket.js
+++ b/tls_socket.js
@@ -240,7 +240,7 @@ function createServer(cb) {
                 socket.emit('error', exception);
             });
 
-            cleartext.on('secureConnect', function() {
+            cleartext.on('secure', function() {
                 log.logdebug('TLS secured.');
                 var cert = cleartext.getPeerCertificate();
                 if (cleartext.getCipher) {

